libfwupdplugin/fu-input-stream.c

```diff
@@ -621,3 +621,74 @@ fu_input_stream_chunkify(GInputStream *stream,
 	}
 	return TRUE;
 }
+
+/**
+ * fu_input_stream_find:
+ * @stream: a #GInputStream
+ * @buf: input buffer to look for
+ * @bufsz: size of @buf
+ * @offset: (nullable): found offset
+ * @error: (nullable): optional return location for an error
+ *
+ * Find a memory buffer within an input stream, without loading the entire stream into a buffer.
+ *
+ * Returns: %TRUE if @buf was found
+ *
+ * Since: 2.0.0
+ **/
+gboolean
+fu_input_stream_find(GInputStream *stream,
+		     const guint8 *buf,
+		     gsize bufsz,
+		     gsize *offset,
+		     GError **error)
+{
+	g_autoptr(GByteArray) buf_acc = g_byte_array_new();
+	const gsize blocksz = 0x10000;
+	gsize offset_add = 0;
+	gsize offset_cur = 0;
+
+	g_return_val_if_fail(G_IS_INPUT_STREAM(stream), FALSE);
+	g_return_val_if_fail(buf != NULL, FALSE);
+	g_return_val_if_fail(bufsz != 0, FALSE);
+	g_return_val_if_fail(bufsz < blocksz, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	while (TRUE) {
```

Comment:
How about as a safety precaution that the loop breaks

`while (offset_cur < bufsz) {`?

Reply:
Makes sense. Fixed!